enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/ArtifactUtils.java

```diff
@@ -18,70 +18,201 @@
+                    .getRoot();
+        } catch (ExpressionEvaluationException | ComponentLookupException e) {
+            throw new EnforcerRuleException("Unable to lookup a component " + e.getLocalizedMessage(), e);
+        } catch (DependencyCollectionException e) {
+            throw new EnforcerRuleException("Could not build dependency tree " + e.getLocalizedMessage(), e);
         }
-        return children;
     }
 
     /**
-     * Checks the set of dependencies against the list of patterns.
+     * <p>Retrieves all <em>child</em> dependency artifacts from the given {@link DependencyNode} and returns them
+     * as a set of {@link Artifact}.</p>
+     * <p><u>Note:</u>&nbsp;Thus, the result will not contain the root artifact.</p>
+     * @param node root node
+     * @return set of all <em>child</em> dependency artifacts
+     */
+    public static Set<Artifact> getDependencyArtifacts(DependencyNode node) {
+        return getDependencyArtifacts(node, new HashSet<>());
+    }
+
+    private static Set<Artifact> getDependencyArtifacts(DependencyNode node, Set<Artifact> set) {
+        node.getChildren().forEach(child -> {
+            set.add(toArtifact(child));
+            getDependencyArtifacts(child, set);
+        });
+        return set;
+    }
+
+    /**
+     * Returns a subset of dependency artifacts that match the given collection of patterns
      *
-     * @param thePatterns the patterns
-     * @param dependencies the dependencies
+     * @param dependencies dependency artifacts to match against patterns
+     * @param patterns patterns to match against the artifacts
      * @return a set containing artifacts matching one of the patterns or <code>null</code>
      * @throws EnforcerRuleException the enforcer rule exception
      */
-    public static Set<Artifact> checkDependencies(Set<Artifact> dependencies, List<String> thePatterns)
+    public static Set<Artifact> filterDependencyArtifacts(Set<Artifact> dependencies, Collection<String> patterns)
             throws EnforcerRuleException {
-        Set<Artifact> foundMatches = null;
-
-        if (thePatterns != null && thePatterns.size() > 0) {
-
-            for (String pattern : thePatterns) {
-                String[] subStrings = pattern.split(":");
-                subStrings = StringUtils.stripAll(subStrings);
-                String resultPattern = StringUtils.join(subStrings, ":");
+        try {
+            return ofNullable(patterns)
+                    .map(collection -> collection.parallelStream()
+                            .map(p -> p.split(":"))
+                            .map(StringUtils::stripAll)
+                            .map(arr -> String.join(":", arr))
+                            .flatMap(pattern -> dependencies.parallelStream()
+                                    .filter(artifact -> compareDependency(pattern, artifact)))
+                            .collect(Collectors.toSet()))
```

Comment:
`parallelStream` and not thread safe `Collectors.toSet()` 😄

Reply:
Yes...